pkg/storage/stores/indexshipper/compactor/deletion/delete_requests_store.go

```diff
@@ -63,60 +64,100 @@ func NewDeleteStore(workingDirectory string, indexStorageClient storage.Client)
+		return nil, err
+	}
 
-	// Add an entry with userID, requestID as range key and status as value to make it easy to manage and lookup status
-	// We don't want to set anything in hash key here since we would want to find delete requests by just status
-	writeBatch := ds.indexClient.NewWriteBatch()
+	return results, nil
+}
+
+func newRequest(req DeleteRequest, requestID []byte, createdAt model.Time, seqNumber int) (DeleteRequest, error) {
+	req.RequestID = string(requestID)
+	req.Status = StatusReceived
+	req.CreatedAt = createdAt
+	req.SequenceNum = int64(seqNumber)
+	if err := req.SetQuery(req.Query); err != nil {
+		return DeleteRequest{}, err
+	}
+	return req, nil
+}
+
+func (ds *deleteRequestsStore) writeDeleteRequest(req DeleteRequest, writeBatch index.WriteBatch) {
+	userIDAndRequestID := deleteRequestHash(req.UserID, req.RequestID, req.SequenceNum)
+
+	// Add an entry with userID, requestID, and sequence number as range key and status as value to make it easy
+	// to manage and lookup status. We don't want to set anything in hash key here since we would want to find
+	// delete requests by just status
 	writeBatch.Add(DeleteRequestsTableName, string(deleteRequestID), []byte(userIDAndRequestID), []byte(StatusReceived))
 
 	// Add another entry with additional details like creation time, time range of delete request and the logQL requests in value
-	rangeValue := fmt.Sprintf("%x:%x:%x", int64(createdAt), int64(startTime), int64(endTime))
-	writeBatch.Add(DeleteRequestsTableName, fmt.Sprintf("%s:%s", deleteRequestDetails, userIDAndRequestID),
-		[]byte(rangeValue), []byte(query))
+	rangeValue := fmt.Sprintf("%x:%x:%x", int64(ds.now()), int64(req.StartTime), int64(req.EndTime))
+	writeBatch.Add(DeleteRequestsTableName, fmt.Sprintf("%s:%s", deleteRequestDetails, userIDAndRequestID), []byte(rangeValue), []byte(req.Query))
 
 	// create a gen number for this result
-	writeBatch.Add(DeleteRequestsTableName, fmt.Sprintf("%s:%s", cacheGenNum, userID), []byte{}, generateCacheGenNumber())
+	writeBatch.Add(DeleteRequestsTableName, fmt.Sprintf("%s:%s", cacheGenNum, req.UserID), []byte{}, generateCacheGenNumber())
+}
 
-	err := ds.indexClient.BatchWrite(ctx, writeBatch)
-	if err != nil {
-		return nil, err
+func deleteRequestHash(userID, requestID string, sequenceNumber int64) string {
+	// Any requests made before the addition of sequence numbers won't have one
+	// Ensure backward compatibility by treating the 0th sequence number as the
+	// old format
+	if sequenceNumber == 0 {
+		return fmt.Sprintf("%s:%s", userID, requestID)
+	}
+	return fmt.Sprintf("%s:%s:%d", userID, requestID, sequenceNumber)
+}
+
+func (ds *deleteRequestsStore) generateID(ctx context.Context, req DeleteRequest) ([]byte, error) {
+	requestID := generateUniqueID(req.UserID, req.Query)
+
+	for {
+		_, err := ds.GetDeleteRequestGroup(ctx, req.UserID, string(requestID))
+		if err != nil {
+			if err == ErrDeleteRequestNotFound {
```

Comment:
hmmm why is `ErrDeleteRequestNotFound` handled differently here?

Reply:
The job of this is to ensure that there are no duplicates. `ErrDeleteRequestNotFound` means we're good to proceed while any other error means that something broke.

Reply:
I've restructured it a little bit in the hope that it is a little clearer.